Minimal HUK implementation without full CAAM driver

Emantor opened the pull request:
This PR implements a minimal implementation to retrieve a Master Key Verification Blob (MKVB) from the CAAM inside of i.MX processors for use as a Hardware Unique Key (HUK). In contrast to the full CAAM implementation offered by NXP, this only implements the necessary bits to retrieve MKVB once, does not allocate job rings to the secure world and has no conflicts with the linux kernel CAAM driver, since the CAAM is not accessed after the MKVB is retrieved.
Things I'm not sure about:
- Access through the structs to the registers:
I followed the style implemented by @bryanodonoghue, it may be more readable to switch to a normal `#define` based register access model

Related:
#2892 - initial i.MX HUK issue
#3149 - full NXP CAAM driver

jforissier commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
+	vaddr_t  ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+
+	uint32_t reg;
+	uint32_t mask;
```
Please initialize

Emantor replied:
Pushed a review commit with the initialized  variables.

jforissier commented on core/arch/arm/plat-imx/imx_caam.h:
```diff
@@ -14,14 +14,23 @@
 	uint32_t			jrmidr_ms;
 	uint32_t			jrmidr_ls;
-};
+} __packed;
```
Should be part of "plat-imx: pack caam structs" I suppose

Emantor replied:
I will fix that while squashing the review commits.

jforissier commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+{
+	struct imx_caam_ctrl *caam = (struct imx_caam_ctrl *)
+		core_mmu_get_va(CAAM_BASE, MEM_AREA_IO_SEC);
+	struct imx_caam_jr *jr;
```
`= NULL;`

jforissier commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+	cache_operation(TEE_CACHEFLUSH, jr->outring,
+			jr->njobs*sizeof(struct imx_caam_outring_entry) + 4);
+
+	// Tell CAAM that one job is available
```
`/* ... */` is preferred

bryanodonoghue commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -50,6 +52,44 @@ static void caam_enable_clocks(bool enable)
+		   virt_to_phys(jr->outring));
+	io_write32((vaddr_t)&caam->jrcfg[0].orsr, jr->njobs);
+	return jr;
+}
```
If I'm reading the code right, you don't need to modify job-ring ownership from non-secure to secure world to do your HUK stuff here, is that correct ?

bryanodonoghue replied:
Follow on question.
What is the effect of this code on runtime Linux ?

Emantor replied:
No, the jobring ownership modification is not necessary since the `PRIBLOB` bits are used to prevent another CAAM user from obtaining the MKVB.

There should be no effect on runtime Linux, since the CAAM is used once during OP-TEE startup and not used afterwards. However this is not the case with the current PR, since `huk_subkey_derive` expects `tee_otp_get_hw_unique_key` to be callable multiple times.
The solution to this is to save the MKVB and only retrieve it once, maybe during an earlier `init_call` which runs before the service init. This needs to be implemented before this PR is ready. (I just dug into this after looking at the system PTA which also uses `huk_subkey_derive`)

Emantor replied:
*sigh* Reading the code further: the `HUK` evidently is not supposed to be retained in memory, since its zeroed explicitly. I guess the attack scenario is someone being able to read the secure memory and thus being able to obtain all encryption keys, but not being able to compomise the `HUK`.

bryanodonoghue replied:
Do you reckon that's a realistic attack ?
From my POV so long as your HUK code doesn't break runtime Linux with job-ring permission conflict it looks fine. So if that means reading the HUK and storing in OPTEE memory my own view is that is not really a problem.

jenswi-linaro replied:
There's a point in not keeping HUK in memory longer than necessary. Because as long as HUK isn't revealed the attacker is limited to tricking OP-TEE into doing unintended things with HUK instead of letting the attacker do whatever he likes with it.

Emantor replied:
I can see the additional security benefit for other platforms, unfortunately this is not possible for the i.MX6 platform AFAICS. We need to increment the PRIBLOB field and can not generate the same MKVB again afterwards.

Emantor commented:
I have a rebased version against the recent master commits and will squash the review commits and suggested changes from @jforissier at the same time on the next push.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
 #include <kernel/generic_boot.h>
 #include <platform_config.h>
 #include <stdint.h>
+#include <mm/core_memprot.h>
```
minor: could shift this before `"include <platform_config.h>`

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
+	else
+		reg &= ~mask;
+
+	io_write32((ccm_base + CCM_CCGR0), reg);
```
minor: can remove inner parentheses, here and at line 48.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
+	struct imx_caam_ctrl *caam = (struct imx_caam_ctrl *)(paddr_t)CAAM_BASE;
 	uint32_t reg;
 	int i;
 
```
suggest `assert(!cpu_mmu_enabled());`

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
+
+	mask = (CCM_CCGR0_CAAM_WRAPPER_IPG  |
+			CCM_CCGR0_CAAM_WRAPPER_ACLK |
+			CCM_CCGR0_CAAM_SECURE_MEM);
```
can remove parentheses.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -5,18 +5,58 @@
+
+static void caam_enable_clocks(bool enable)
+{
+	vaddr_t  ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
```
minor: 2 spaces

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.h:
```diff
@@ -10,6 +10,7 @@
 
 #include <imx-regs.h>
 #include <stdint.h>
+#include <compiler.h>
```
minor: move above `#include <ims-regs.h>`

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.h:
```diff
@@ -9,6 +9,7 @@
 #define __IMX_CAAM_H__
 
 #include <imx-regs.h>
+#include <types_ext.h>
```
minor: move below `#include <stdint.h>`

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.h:
```diff
@@ -32,13 +33,53 @@ struct imx_caam_job_ring {
+	uint32_t	jrcr;
+	uint8_t		padding12[1688];
+	uint32_t	jraav;
+	uint8_t		padding13[2296];
```
Cross check the structure. I could 4100 bytes, i would guess it is 4096 bytes aligned.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -50,6 +52,44 @@ static void caam_enable_clocks(bool enable)
+	} while ((reg_val & 0x1) && --timeout);
+}
+
+static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
```
curerntly unused. maybe add `_unused` and remove it in next commit.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+
+	cache_operation(TEE_CACHEINVALIDATE, jr->outring,
+			sizeof(struct imx_caam_outring_entry));
+	cache_operation(TEE_CACHEINVALIDATE, &outbuf, MKVB_SIZE);
```
This sequence invalidates cache lines related to the stack and to the heap. Hazardous regarding buffer alignment.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+		counter++;
+		if (counter > 1000)
+			break;
+	}
```
consider using `timeout_elpased()` here rather than ISBs/DSBs.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+	io_write32((vaddr_t)&caam->scfgr, PRIBLOB_11);
+	free(jr->inring);
+	free(jr->outring);
+	jr->ctrl = NULL;
```
minor: `jr->ctrl = NULL;` not needed.

etienne-lms commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -90,6 +90,67 @@ static struct imx_caam_jr *caam_init_jr(struct imx_caam_ctrl *caam)
+	descriptor[2] = virt_to_phys(&outbuf);
+	cache_operation(TEE_CACHEFLUSH, &outbuf, MKVB_SIZE);
+
+	cache_operation(TEE_CACHEFLUSH, &descriptor, 16);
```
s/16/sizeof(descriptor)/

Emantor commented on core/arch/arm/plat-imx/imx_caam.c:
```diff
@@ -50,6 +52,44 @@ static void caam_enable_clocks(bool enable)
+		dsb();
+		reg_val = io_read32((vaddr_t)&caam->jrcfg[0].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val == 0x4) && --timeout);
```
`timeout_init_us` and `timeout_elapsed` won't work AFAICS since they use the generic timer which is not available on Cortex-A9 (i.MX6 Q/D/L/S)

etienne-lms replied:
you're right. sorry. bad idea.

Emantor commented:
I have rebased against latest master and changed to stack allocation. With the alignement within the structure to 64 bytes we should always flush and invalidate the intended date, without hitting stack data above or below. I also tried to incorporated all of the review comments, Thanks @jforissier, @etienne-lms, @jenswi-linaro & @bryanodonoghue. @etienne-lms I did not apply your Ack since I think the code has changed enough to require another review.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -13,12 +13,17 @@
 #include "imx_caam.h"
 
+uint8_t stored_key[32] = { 0 };
+bool mkvb_retrieved;
```
`static` bot both. No need to init `stored_hey[]`.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -53,6 +58,103 @@ static void caam_enable_clocks(bool enable)
+	while (io_read32((vaddr_t)&mkvb.ctrl->jrcfg[0].orsfr) != 1) {
+		counter++;
+		if (counter > 10000)
+			break;
```
break or panic or goto out?

cneveux commented:
@Emantor 

I understand that you want to have the HUK generated with the CAAM but I don't see the interest of bypassing all the work we, NXP, are doing to enable the CAAM driver. In your PR you have to enable the CAAM anyway to use the Job Ring. In plus you have missed a condition that is really needed to be full secure. This is the selection of the master key. 

In the driver we want to add and that has been improved and validated on all i.MX devices (taking care of all boot modes). We are doing the same in a more generic way.

Could you explain us what is the added value of your implementation that is a kind of copy of what has been submitted in the PR #3149?

Emantor commented:
> I understand that you want to have the HUK generated with the CAAM but I don't see the interest of bypassing all the work we, NXP, are doing to enable the CAAM driver. In your PR you have to enable the CAAM anyway to use the Job Ring.

I am interested in generating the MKVB for use as a HUK. I am not interested in a CAAM driver implementation, since I don't see the use case for it. The CAAM driver adds a lot of code which increases the TCB for my secure OS. For our use cases a full CAAM driver within OP-TEE is not needed, since we don't require hardware acceleration for the crypto primitives.

> In plus you have missed a condition that is really needed to be full secure. This is the selection of the master key.

According to the documentation at hand, the `OTPMK` is used as the default Master Key CAAM input.
I am well aware that the OTPMK is not available on none HAB-enabled devices and a test key is used.
I'd argue that it does not make much sense to deploy devices without HAB and with OP-TEE.
> 
> In the driver we want to add and that has been improved and validated on all i.MX devices (taking care of all boot modes). We are doing the same in a more generic way.
Which is also currently incompatible to the mainline kernel due to possible clock problems. If somebody i.e. decides to add power management to the Linux CAAM driver, OP-TEE will be stuck when trying to access the CAAM. I agree that a solution to the problem exists which is to move the clocks into the secure world. However this requires discussions with the upstream kernel maintainers on how this implementation is going to work.
> Could you explain us what is the added value of your implementation that is a kind of copy of what has been submitted in the PR #3149?
- It adds the required minimum amount of code to derive a HUK for OP-TEE on bootup without requiring a switch of the full crypto implementations to a hardware devices.
- It allows usage of all CAAM jobrings within Linux
- The CAAM is not touched after OP-TEE initialization, which eliminates all problems stemming from OP-TEE requiring the CAAM clock to stay enabled.
- This code can later be refactored to take advantage of the real implementations provided in the full CAAM driver.

ricardosalveti commented:
I also share similar concerns and needs with @Emantor, so I'm also interested at having just the minimum to derive a HUK instead of having the full CAAM support in place.

Emantor commented:
@dmcilvaney this may also be relevant to your interests for the Microsoft IOT devices.

cneveux commented:
@ricardosalveti, @Emantor 
Well, thinks is that your are not taking care of all devices and all devices boot mode.
You can not hardcode the number of Job Rings, number is function of device type. You can not assume that JR#0 is always available for the OPTEE at boot time. This is function of the boot mode if there is a previous SW doing the authentication of booting OSes first.
Next we (NXP) are doing the implementation of the CAAM driver to satisfy everybody. This driver is working since long time and this is why we are upstreaming it now. The driver is  not so complex and no heavy, the driver is doing the minimum if you don't enable the other crypto algorithm. 
What @Emantor did, is just copying the work I submitted and re-write it. If you want to keep this version feel free to keep in your own source tree but please let us push our work that is expected by many customers (directly or not).

Thanks for you understanding.

Emantor commented:
> @ricardosalveti, @Emantor
> Well, thinks is that your are not taking care of all devices and all devices boot mode.
> You can not hardcode the number of Job Rings, number is function of device type. You can not assume that JR#0 is always available for the OPTEE at boot time. This is function of the boot mode if there is a previous SW doing the authentication of booting OSes first.

I tried to find some documentation on which job ring is actually used by the HAB, but all I could find were discussion on the [LKML](https://lore.kernel.org/patchwork/cover/877819/) and [U-Boot](https://marc.info/?t=155448108300001&r=1&w=2). Do you have more documentation internally that you could share?

To be clear: this HUK patch breaks the case where your boot chain is `SPL`->`OP-TEE`->`U-Boot`->`kernel` and U-Boot tries to use HAB to authenticate additional images. Are there similar restrictions on the other job rings? Is the simple solution to switch to job ring 2?

This is a bug in this PR and should consequently be fixed.
> Next we (NXP) are doing the implementation of the CAAM driver to satisfy everybody. This driver is working since long time and this is why we are upstreaming it now. The driver is not so complex and no heavy, the driver is doing the minimum if you don't enable the other crypto algorithm.

With a diffstat of `+13,804 −62` for the current CAAM HUK patch set I'd call this a lot more complex :-). Of course not all code paths are used, but getting the whole PR review takes time. As an alternative for now I posted this PR.
It also requires currently undocumented device tree changes, i.e. `secure-status: okay` for the CAAM ctrl node and a job ring.
And your PR notes that it currently can't be used without other PRs which have not been posted yet and makes it quite hard to judge whether the HUK derivation works correctly. IMO the HUK PR can only be reviewed in conjunction with the SNVS patches.
> What @Emantor did, is just copying the work I submitted and re-write it. If you want to keep this version feel free to keep in your own source tree but please let us push our work that is expected by many customers (directly or not).

I'm not restricting you ability to push your changes upstream in any way. I'm just proposing a solution for now which is easier to review and test and has no immediate impact on the kernel.

Emantor commented:
@MrVan any comments on this patch series? This requires your `Reviewed-by` because of the `plat-imx` changes.

Emantor commented:
I pushed a patch which switches to using job ring two, which is the same ring used as by the full caam implementation for the `CFG_CAAM_DT=n` case.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -53,6 +58,103 @@ static void caam_enable_clocks(bool enable)
+	do {
+		reg_val = io_read32((vaddr_t)&ctrl->jrcfg[0].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val == 0x4) && --timeout);
```
what is timeout elapses? panic?
If not, maybe timeout test is useless.

Emantor replied:
I'm returning `TEE_ERROR_SECURITY` now.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -53,6 +58,103 @@ static void caam_enable_clocks(bool enable)
+	mkvb.ctrl = (struct imx_caam_ctrl *)
+		core_mmu_get_va(CAAM_BASE, MEM_AREA_IO_SEC);
+
+	caam_enable_clocks(true);
```
caam clocks were already enabled from `init_caam()`.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -53,6 +58,103 @@ static void caam_enable_clocks(bool enable)
+	cache_operation(TEE_CACHEFLUSH, &mkvb.jr,
+			sizeof(mkvb.jr.inring[0]));
+
+	/*  Tell CAAM that one job is available */
```
minor: remove 1 space here and at line 123.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.h:
```diff
@@ -34,6 +34,14 @@ struct imx_caam_job_ring {
+#define MKVB_DESC_BLOB			0x870d0002
+
+/* PRIBLOB Bits */
+#define PRIBLOB_11			3
```
add an empty space below.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,152 @@
 
 #include "imx_caam.h"
 
+static uint8_t stored_key[32] = { 0 };
```
suggest to replace `32` with `MKVB_SIZE` for consistency with line 137.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,152 @@
+
+		io_write32(ccm_base + CCM_CCGR6, reg);
+	}
+
```
remove empty line

Emantor replied:
Done.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -63,15 +63,15 @@ static void imx_caam_reset_jr(struct imx_caam_ctrl *ctrl)
 	uint32_t timeout = 1000;
 
-	io_write32((vaddr_t)&ctrl->jrcfg[0].jrcr, 1);
+	io_write32((vaddr_t)&ctrl->jrcfg[1].jrcr, 1);
```
Suggest to use a macro, a global or a argument for identifying the job ring index.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -62,26 +62,40 @@ static void imx_caam_reset_jr(struct imx_caam_ctrl *ctrl)
+static TEE_Result mkvb_init_jr(struct imx_mkvb *mkvb)
 {
 	struct imx_caam_ctrl *ctrl = mkvb->ctrl;
+	TEE_Result ret;
```
minor: default init to something

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -109,10 +125,10 @@ static TEE_Result caam_get_mkvb(uint8_t *dest)
 	io_write32((vaddr_t)&mkvb.ctrl->jrcfg[MKVB_JR].irjar, 1);
 
-	/*  Busy loop until job is completed */
+	/* Busy loop until job is completed */
```
must reset `ret` to an error code (`ret = TEE_ERROR_SECURITY`) since storage of `mkvb_init_jr()` return code.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -21,10 +21,10 @@
 #include "imx_caam.h"
 
-static uint8_t stored_key[32] = { 0 };
+static uint8_t stored_key[MKVB_SIZE] = { 0 };
```
minor: no need to explicitly init this.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -36,10 +36,7 @@ static void caam_enable_clocks(bool enable)
-		reg |= mask;
-	else
-		reg &= ~mask;
+	reg |= mask;
```
suggest:
```
static void caam_enable_clocks(void)
{
	vaddr_t ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);

	io_setbits32(ccm_base + CCM_CCGR0,
		     CCM_CCGR0_CAAM_WRAPPER_IPG  |
		     CCM_CCGR0_CAAM_WRAPPER_ACLK |
		     CCM_CCGR0_CAAM_SECURE_MEM);

	if (soc_is_imx6dqp() || soc_is_imx6sdl() || soc_is_imx6dq())
		io_setbits32(ccm_base + CCM_CCGR6,
			     CCM_CCGR6_EMI_SLOW);
}
```

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -82,13 +75,13 @@ static void mkvb_init_jr(struct imx_mkvb *mkvb)
 	mkvb->njobs = 4;
-	io_write32((vaddr_t)&ctrl->jrstartr, 1);
-	io_write32((vaddr_t)&ctrl->jrcfg[1].irbar_ls,
+	io_write32((vaddr_t)&ctrl->jrstartr, 2);
```
This change `1` into `2`. Please crosscheck it is what you expect.

Emantor replied:
Yes, this was a bug previously, luckily the bootloader started the second job ring as well as the first. Thanks for noticing :-)

Emantor commented:
@etienne-lms, @clementfaure any comments on this? This is a valid minimal implementation which does not require the whole CAAM stack as implemented by NXP, but can be refactored as the necessary NXP commits are added.

etienne-lms commented:
@Emantor, https://github.com/OP-TEE/optee_os/pull/3228 is not far from landing. Hence maybe consider refactoring on the P-R to plug HUK over the proposed CAAM support. Should find a way to use only features you need from it.

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,148 @@
+	if (mkvb.jr.outring[0].status != 0)
+		ret = TEE_ERROR_SECURITY;
+		goto out;
+
```
Missing { } otherwise it always ends up going to 'goto out'.

Emantor replied:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

Emantor replied:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

Emantor replied:
Yes, this is already addressed in my local branch. I plan to rebase onto the CAAM work, but that might take a bit of refactoring to correctly disable the CAAM driver after HUK generation.

ricardosalveti replied:
Great, thanks for the heads up.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,148 @@
+	uint32_t reg_val = 0;
+	uint32_t timeout = 1000;
+
+	io_write32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrcr, 1);
```
It would be nice that raw value (`1` here) at lines 50, 55, 59, ... are replaced with macros, i.e. `CAAM_JRCR_xxx BIT(0)`, `CAAM_JRINTR_STATE_MASK  GENMASK_32(3, 2)`, ...

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,148 @@
+
+static TEE_Result caam_get_mkvb(uint8_t *dest)
+{
+	struct imx_mkvb mkvb = { 0 };
```
minor: prefer `{ }`

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,148 @@
+			return ret;
+		mkvb_retrieved = true;
+	}
+	memcpy(&hwkey->data, &stored_key, sizeof(hwkey->data));
```
prefer add empty line after `if() { }` block and before last `return`.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -25,6 +164,7 @@ static TEE_Result init_caam(void)
 	if (!caam)
 		return TEE_ERROR_GENERIC;
 
+	caam_enable_clocks();
```
add an empty line below.

etienne-lms commented on core/arch/arm/plat-imx/drivers/imx_caam.c:
```diff
@@ -9,9 +9,148 @@
+	do {
+		reg_val = io_read32((vaddr_t)&ctrl->jrcfg[MKVB_JR].jrintr);
+		reg_val &= 0xc;
+	} while ((reg_val & 0x1) && --timeout);
```
`1` does not seems the expect value, since mask at line 58.
